codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -401,6 +416,17 @@ public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers
                                 "Multiple content-length headers received");
                     }
                 }
+            } else if (validateHeaders && !headers.isEmpty()) {
+                // Need to check trailers don't contain pseudo headers. According to RFC 9113
+                // Trailers MUST NOT include pseudo-header fields (Section 8.3).
+                for (Iterator<Entry<CharSequence, CharSequence>> iterator =
+                    headers.iterator(); iterator.hasNext();) {
```

Comment:
@ejona86 @pkoenig10 The usage of `isEmpty` was fixed in https://github.com/grpc/grpc-java/pull/10663 and #13717, but we did not look closely enough at the problem, because on this line we also call the `iterator()` method on the `Http2Headers`, which `AbstractHttp2Headers` still does not implement: https://github.com/grpc/grpc-java/blob/75492c8b36c9392d94dfce09fc132ce88c94325e/netty/src/main/java/io/grpc/netty/AbstractHttp2Headers.java#L499

Reply:
I see there's already a grpc-java issue on this: https://github.com/grpc/grpc-java/issues/10837

Reply:
This looks like it was supposed to be fixed in 1.55.0 with https://github.com/grpc/grpc-java/pull/9979. The report I got was using 1.51.1, and https://github.com/grpc/grpc-java/issues/10837 reports using 1.30.2, so this is probably not an issue for those who are on newer versions!

Reply:
Oh that's great to hear. I saw the issue this morning as felt, "oh no." Glad it is already added. (Although obviously, better than recently being added would have been "always had been implemented"... We should re-run the benchmarks to see how much of a performance difference it makes to use our own impl, although benchmarking garbage creation is _terribly_ difficult to get meaningful results.)